### PR TITLE
Learning rate schedule how-to

### LINKED ISSUE #222: Learning rate schedule HOWTO
Is there interest in a PR showing how to use learning rate schedules?

### PR DESCRIPTION
**Pull request by joaogui1**
Closes #222 
Reviews and suggestions appreciated!

### CODE REVIEW

File: docs/howtos/lr_schedule.rst [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      loss = cross_entropy_loss(logits, batch['label'])
+      return loss, logits
+    
+    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
+    (_, logits), grad = grad_fn(optimizer.target)
+    step = optimizer.state.step
+    lr = learning_rate_fn(step)
+    optimizer = optimizer.apply_gradient(grad, {"learning_rate":lr})
+    metrics = compute_metrics(logits, batch['label'])
+    return optimizer, metrics
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:43]: This is a great use-case for our brand new Sphinx codediff directive! It allows you to display code in side-by-side view. Can you check how this looks if you compile it locally?

```suggestion
.. codediff::
  :title_left: Default learning rate
  :title_right: Triangular learning rate schedule
  
  @jax.jit
  def train_step(optimizer, batch): #!
    def loss_fn(params):
      logits = CNN().apply({'params': params}, batch['image'])
      loss = cross_entropy_loss(logits, batch['label'])
      return loss, logits
    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
    (_, logits), grad = grad_fn(optimizer.target)


    optimizer = optimizer.apply_gradient(grad) #!
    metrics = compute_metrics(logits, batch['label'])
    return optimizer, metrics
  ---
  @jax.jit
  def train_step(optimizer, batch, learning_rate_fn): #!
    def loss_fn(params):
      logits = CNN().apply({'params': params}, batch['image'])
      loss = cross_entropy_loss(logits, batch['label'])
      return loss, logits
    grad_fn = jax.value_and_grad(loss_fn, has_aux=True)
    (_, logits), grad = grad_fn(optimizer.target)
    step = optimizer.state.step #!
    lr = learning_rate_fn(step) #!
    optimizer = optimizer.apply_gradient(grad, {"learning_rate":lr}) #!
    metrics = compute_metrics(logits, batch['label'])
    return optimizer, metrics
```

File: docs/howtos/lr_schedule.rst [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+And the train_epoch function:
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:45]: Could you rewrite this codeblock using the codediff directive as well, as I did above?

File: docs/howtos/lr_schedule.rst [hunk 3]
```diff
@@ -0,0 +1,74 @@
+Managing Parameters and State
+=============================
+
+We will show you how to...
+
+* define a learning rate schedule
+* train a simple model using that schedule
+
+.. code-block:: python
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:9]: We are switching to the `testcode` directive instead of `code-block`. This will run the code as part of our tests and make sure everything still compiles.

```suggestion
.. testcode::
```

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Thanks a lot for adding this, very exciting!

By the way, did you know you can compile the docs locally very easily? Just go to the docs folder and run `make html`, which compiles them in the `docs/_build/html` folder. I find that very useful for debugging (especially when using code).

If you knew this already then please ignore this comment!

File: docs/howtos/lr_schedule.rst (hunk 1 above)
Comment (joaogui1) [reply docs/howtos/lr_schedule.rst:43]: It looks good, though do you know if there's a way to align the top of the code-boxes?

File: docs/howtos/lr_schedule.rst (hunk 2 above)
Comment (joaogui1) [reply docs/howtos/lr_schedule.rst:45]: Is there a way to get syntax highlighting with codediff?

Comment (marcvanzee) [reply docs/howtos/lr_schedule.rst:65]: Yes, you simply append #! to each line you want to highlight.

File: docs/howtos/lr_schedule.rst (hunk 1 above)
Comment (marcvanzee) [reply docs/howtos/lr_schedule.rst]: They should be align to the top by default. Did you sync to head? I made some changes to the CSS recently so many they aren't yet part of your change?

Comment (marcvanzee) [reply docs/howtos/lr_schedule.rst]: Ah sorry I see those changes are not merged yet, they are actually part of #985 . I'll split off those CSS changes and get them merged so you can sync to head.

Comment (marcvanzee) [reply docs/howtos/lr_schedule.rst]: I just merged #990, could you sync to head and let me know how it looks now?

Comment (joaogui1) [reply docs/howtos/lr_schedule.rst]: Looks great now!

**Comment by joaogui1** [comment]
Any other changes?

File: docs/howtos/lr_schedule.rst [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+And the train_epoch function:
+
+.. codediff::
+  :title_left: Default learning rate
+  :title_right: Triangular learning rate schedule
+  
+  def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
+  """Train for a single epoch."""
+  train_ds_size = len(train_ds['image'])
+  steps_per_epoch = train_ds_size // batch_size
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:74]:
```suggestion
  steps_per_epoch = train_ds_size // batch_size
  
  
  
```

File: docs/howtos/lr_schedule.rst [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  # compute mean of metrics across each batch in epoch.
+  batch_metrics_np = jax.device_get(batch_metrics)
+  epoch_metrics_np = {
+      k: np.mean([metrics[k] for metrics in batch_metrics_np])
+      for k in batch_metrics_np[0]}
+
+  logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
+               epoch_metrics_np['loss'], epoch_metrics_np['accuracy'] * 100)
+
+  return optimizer, epoch_metrics_np
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:94]: Few suggestions to make the code less wide so it looks better in the side-by-side diff.

```suggestion
  perms = perms[:steps_per_epoch * batch_size]
  perms = perms.reshape((steps_per_epoch, batch_size))
  batch_metrics = []
  for perm in perms:
    batch = {k: v[perm, ...] for k, v in train_ds.items()}
    optimizer, metrics = train_step(optimizer, batch) #!
    batch_metrics.append(metrics)

  # compute mean of metrics across each batch in epoch.
  batch_metrics = jax.device_get(batch_metrics)
  epoch_metrics = {
      k: np.mean([metrics[k] for metrics in batch_metrics])
      for k in batch_metrics[0]}

  logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
               epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)

  return optimizer, epoch_metrics
```

File: docs/howtos/lr_schedule.rst [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    # compute mean of metrics across each batch in epoch.
+    batch_metrics_np = jax.device_get(batch_metrics)
+    epoch_metrics_np = {
+        k: np.mean([metrics[k] for metrics in batch_metrics_np])
+        for k in batch_metrics_np[0]}
+
+    logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
+                epoch_metrics_np['loss'], epoch_metrics_np['accuracy'] * 100)
+
+    return optimizer, epoch_metrics_np
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:121]: Same here

```suggestion
    # If you want 4 cycles per epoch. #!
    learning_rate_fn = create_triangular_schedule(3e-3, 3e-2, #!
                                                  steps_per_epoch//4) #!
    perms = jax.random.permutation(rng, len(train_ds['image']))
    perms = perms[:steps_per_epoch * batch_size]
    perms = perms.reshape((steps_per_epoch, batch_size))
    batch_metrics = []
    for perm in perms:
      batch = {k: v[perm, ...] for k, v in train_ds.items()}
      optimizer, metrics = train_step(optimizer, batch, #!
                                      learning_rate_fn) #!
      batch_metrics.append(metrics)

    # compute mean of metrics across each batch in epoch.
    batch_metrics = jax.device_get(batch_metrics)
    epoch_metrics = {
        k: np.mean([metrics[k] for metrics in batch_metrics])
        for k in batch_metrics[0]}

    logging.info('train epoch: %d, loss: %.4f, accuracy: %.2f', epoch,
                epoch_metrics['loss'], epoch_metrics['accuracy'] * 100)

    return optimizer, epoch_metrics
```

File: docs/howtos/lr_schedule.rst [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+  return optimizer, epoch_metrics_np
+  ---
+  def train_epoch(optimizer, train_ds, batch_size, epoch, rng):
+    """Train for a single epoch."""
+    train_ds_size = len(train_ds['image'])
+    steps_per_epoch = train_ds_size // batch_size
+
+    #If you want 4 cycles per epoch #!
+    learning_rate_fn = create_triangular_schedule(3e-3, 3e-2, steps_per_epoch//4) #!
```
Comment (marcvanzee) [inline docs/howtos/lr_schedule.rst:102]: Is there anything we can say to motivate this choice of hyperparameters? How did you choose them? Are these the standard values that people usually start with?

**Review by marcvanzee — CHANGES_REQUESTED** [review]
Looks really awesome already!

I added a bunch of comments, mostly improving some formatting.

**Comment by joaogui1** [comment]
I added all of your suggestions, thank you very much!

File: docs/howtos/lr_schedule.rst (hunk 7 above)
Comment (marcvanzee) [reply docs/howtos/lr_schedule.rst:115]: Let's leave this for now so we can get the PR merged, since it is not super important anyway.

**Review by marcvanzee — APPROVED** [review]

**Comment by marcvanzee** [comment]
Awesome, I merged it! Thanks a lot for this cool contribution 👍
